packages/notte-browser/src/notte_browser/session.py

```
@@ -357,46 +386,150 @@ async def locate(self, action: BaseAction) -> Locator | None:
-        Execute an action, either by passing a BaseAction as the first argument, or by passing ExecutionRequestDict fields as kwargs.
+        Execute an action, either by passing a BaseAction as the first argument, or by passing action fields as kwargs.
         """
-        step_action = ExecutionRequest.get_action(action=action, data=data)
+        # Fast path: if action is already a BaseAction, use it directly
+        if isinstance(action, BaseAction):
+            step_action = action
+        elif kwargs:
+            # Convert kwargs to BaseAction using fast mapping
+            step_action = action_dict_to_base_action(kwargs)  # type: ignore[arg-type]
+        elif action is None:
+            raise ValueError("No action provided")
```

Ruff (0.14.8)

Avoid specifying long messages outside the exception class

(TRY003)